### ISSUE TITLE
Font rendering issues (strange sizes / spacing)

### ISSUE THREAD

**Issue by fdev31** [kovidgoyal/kitty#1376]
Kitty doesn't seem to render my fonts properly, it's the only app where I see those issues.
I'm currently using the Fira Code font, ligatures and rendering is OK using the Geany editor.
Also I can make a side by side comparison with another terminal emulator (Terminator), which doesn't support ligatures but renders OK:
 
![shot](https://user-images.githubusercontent.com/238622/52735586-43dc8180-2fc8-11e9-99c8-a8713830281a.png)

I tried all combinations of otf & ttf fonts, currently I have both installed.
I can see 3 different issues:
- spacing, much less text is rendered on screen
- icons are rendered too small
- colors are not matching the expected set, as if I didn't have a 256 color enabled terminal
   - I tried to set `term xterm-256color` without getting any visual change

kitty --debug-config:                                                                                                                                                                      
```
kitty 0.13.3 created by Kovid Goyal
Linux popo 4.20.6-arch1-1-ARCH #1 SMP PREEMPT Thu Jan 31 08:22:01 UTC 2019 x86_64
Running under: X11
Arch Linux \r (\l)
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
Loaded config files: /home/fab/.config/kitty/kitty.conf

Config options different from defaults:
background_opacity 0.9
font_family        Fira Code
font_size          12.0
```
kitty --debug-font-fallback:
```
[044 19:46:14.962098] Preloaded font faces:
[044 19:46:14.962135] normal face: /usr/share/fonts/OTF/FiraCode-Regular.otf:0
[044 19:46:14.962152] bold face: /usr/share/fonts/TTF/FiraCode-Bold.ttf:0
[044 19:46:14.962166] italic face: /usr/share/fonts/OTF/FiraCode-Regular.otf:0
[044 19:46:14.962183] bi face: /usr/share/fonts/TTF/FiraCode-Bold.ttf:0
[044 19:46:15.482400] [PARSE ERROR] Unknown OSC code: 7
U+f0c3 Face(family=Hack Nerd Font Mono, style=Regular, ps_name=HackNerdFontCompleteM-Regular, path=/usr/share/fonts/TTF/Hack-Regular-Nerd-Font-Complete-Mono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-355, underline_thickness=90) (new face)
U+f126 Face(family=Hack Nerd Font Mono, style=Regular, ps_name=HackNerdFontCompleteM-Regular, path=/usr/share/fonts/TTF/Hack-Regular-Nerd-Font-Complete-Mono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-355, underline_thickness=90)
U+f06a Face(family=Hack Nerd Font Mono, style=Regular, ps_name=HackNerdFontCompleteM-Regular, path=/usr/share/fonts/TTF/Hack-Regular-Nerd-Font-Complete-Mono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-355, underline_thickness=90)
U+f015 Face(family=Hack Nerd Font Mono, style=Regular, ps_name=HackNerdFontCompleteM-Regular, path=/usr/share/fonts/TTF/Hack-Regular-Nerd-Font-Complete-Mono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-355, underline_thickness=90)
U+2714 Face(family=DejaVu Sans Mono, style=Book, ps_name=DejaVuSansMono, path=/usr/share/fonts/TTF/DejaVuSansMono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-175, underline_thickness=90) (new face)
U+f017 Face(family=Hack Nerd Font Mono, style=Regular, ps_name=HackNerdFontCompleteM-Regular, path=/usr/share/fonts/TTF/Hack-Regular-Nerd-Font-Complete-Mono.ttf, index=0, is_scalable=True, has_color=False, ascender=1901, descender=-483, height=2384, underline_position=-355, underline_thickness=90)
```

kitty list-fonts (now showing everything twice since otf+ttf are installed, but it didn't seem to make any difference):
```
Andale Mono
    Andale Mono

Courier New
    Courier New
    Courier New Bold
    Courier New Bold Italic
    Courier New Italic

DejaVu Sans Mono
    DejaVu Sans Mono
    DejaVu Sans Mono Bold
    DejaVu Sans Mono Bold Oblique
    DejaVu Sans Mono Oblique

Droid Sans Mono
    Droid Sans Mono

Fira Code
    Fira Code Bold
    Fira Code Bold
    Fira Code Light
    Fira Code Light
    Fira Code Medium
    Fira Code Medium
    Fira Code Regular
    Fira Code Regular
    Fira Code Retina
    Fira Code Retina

FreeMono
    FreeMono
    FreeMono Bold
    FreeMono Bold Oblique
    FreeMono Oblique

Hack
    Hack Bold
    Hack Bold Italic
    Hack Italic
    Hack Regular

Inconsolata
    Inconsolata Bold
    Inconsolata Regular

MathJax_Vector
    MathJax_Vector-Regular

MathJax_Vector-Bold
    MathJax_Vector-Bold

Nimbus Mono PS
    NimbusMonoPS-Bold
    NimbusMonoPS-BoldItalic
    NimbusMonoPS-Italic
    NimbusMonoPS-Regular

```

**Comment by Luflosi** [comment]
The issue with the different colours can probably be solved by changing the 16 colours in your `kitty.conf` to be the same as in Terminator, see https://sw.kovidgoyal.net/kitty/conf.html#the-color-table.

**Comment by fdev31** [comment]
Thank you for the fast feedback, I figured that for colors that was related to the default scheme.... sorry for this part of the report!
For the color spacing, I can mitigate the problem with adjust_column_width -2, it's not really perfectly rendering but it's more compact at least...

**Comment by fdev31** [comment]
![small](https://user-images.githubusercontent.com/238622/52741282-4b565780-2fd5-11e9-9c25-18b28b7fae8f.png)

I just saw this issue... when I use the "adjust_column_width -2" workaround.
playing a bit more I figured that it is related to the adjust_column_width value, if I make it -5 and then scale the font up I get cartoonish results on some texts:

![more](https://user-images.githubusercontent.com/238622/52741484-b738c000-2fd5-11e9-992a-b97922799b6e.png)

**Comment by kovidgoyal** [comment]
I dont see anything wrong with spacing, that is the amount of space this font requires. kitty unlike most terminal emulators actually renders text in cells. Which means the cell has to be as wide as the widest character in the font. If you make the cell narrower, some characters wont fit, and wil be resized to fit in the adjusted cell, as you discovered.

As for your icons being too small, it depends on what you are using for the oicons. If you are using Nerd fonts, then follow them by space characters. And you will be fine.

**Comment by fdev31** [comment]
I'm not sure to get your point, you don't see anything wrong in my screenshots ?
As well, there is a space after each character "by default".

I believe that relative scale of characters should be preserved, to avoid "cartoon" rendering, it really looks like an issue to me...

Look at the text "normal" the small "a" doesn't take the whole width and there is plenty of space on the top....

**Comment by kovidgoyal** [comment]
In a monospace font, all characters have to be rendered in a cell of the same width. That is the nature of a monospace font. That means that there will be extra space around an i or an a but not a w or M.

**Comment by Luflosi** [comment]
I think with "cartoon" rendering he means that in his last two screenshots some characters are less tall than others, for example the "A" in "NORMAL" and the "S", "R" and "T" of "INSERT".

**Comment by kovidgoyal** [comment]
That will happen if you adjust colum width to less than needed for those
characters.

**Comment by fdev31** [comment]
Indeed this issue is related to the value I mentioned.
I am not exactly sure for the extra spaces thing, but for the issue related to small logos I finally fixed it by removing the nerd font package I had on my system. Not sure why this font was involved even though the glyph still shows (and with correct size) without it - maybe a mention in the FAQ could help someone else.
Anyway thank you for this project, I just switched my default VT ;)
